File: 10_Binary_search_2D_array/search_row_col_sorted_arr.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
  public:
    bool matSearch(vector<vector<int>> &mat, int x) {
        int n = mat.size(), m = mat[0].size();
        int row = 0, col = m - 1;

        while (row < n && col >= 0) {
            if (mat[row][col] == x) return true;
            else if (mat[row][col] < x) row++;
            else col--;
        }
        return false;
    }
};

int main() {
    Solution sol;

    // Sample matrix
    vector<vector<int>> mat = {
        {10, 20, 30, 40},
        {15, 25, 35, 45},
        {27, 29, 37, 48},
        {32, 33, 39, 50}
    };

    // Test case 1
    int x = 29;
    bool result = sol.matSearch(mat, x);
    if (result) {
        cout << "Element " << x << " found in the matrix." << endl;
    } else {
        cout << "Element " << x << " not found in the matrix." << endl;
    }

    // Test case 2
    x = 100;
    result = sol.matSearch(mat, x);
    if (result) {
        cout << "Element " << x << " found in the matrix." << endl;
    } else {
        cout << "Element " << x << " not found in the matrix." << endl;
    }

    return 0;
}

File: 10_Binary_search_2D_array/search_row_col_sorted_arr.java
public class search_row_col_sorted_arr {
    public static boolean matSearch(int mat[][], int x) {
        // your code here
        int n = mat.length, m = mat[0].length;
        int row = 0, col = m - 1;

        while (row < n && col >= 0) {
            if (mat[row][col] == x) {
                return true; 
            }else if (mat[row][col] < x) {
                row++; 
            }else {
                col--;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        // Sample matrix
        int[][] mat = {
            {10, 20, 30, 40},
            {15, 25, 35, 45},
            {27, 29, 37, 48},
            {32, 33, 39, 50}
        };

        // Test values to search for
        int x = 29;
        boolean result = matSearch(mat, x);

        // Output the result
        if (result) {
            System.out.println("Element " + x + " found in the matrix.");
        } else {
            System.out.println("Element " + x + " not found in the matrix.");
        }

        // Test with another value
        x = 100;
        result = matSearch(mat, x);

        if (result) {
            System.out.println("Element " + x + " found in the matrix.");
        } else {
            System.out.println("Element " + x + " not found in the matrix.");
        }
    }
}
